### ISSUE TITLE
Q: additional special record handling capable?

### ISSUE THREAD

**Issue by DarkonIV** [JoshClose/CsvHelper#942]
Can csvHelper work in the situation where the CSV file spec is:

**VersionRec (optional) = @ version x.y.z**
- _MUST be the first record in the file, when version info required by consumer process._
- _MUST begin with @ symbol in first byte of record._
- _MUST have " version " text tag to indicate the type of value to follow as a version._
- _MUST have a version value in #, #.#, or #.#.# format._
- _Our CSV file spec allows for other @ meta tags to be defined in the future which are to be used in aiding the processing of the CSV data to follow in order to provide dynamic CSV updates while maintaining processing capability of multiple schema revisions._

**HeaderRec (required) = # c1_hdr, c2_hdr, c3_hdr**
- _MUST be present in the file when one or more data records present._
- _MUST begin with # in first byte of record._
- _MUST be written upon creation of the first data record by the producing application._

**CommentRec (optional) = # any info as comment to follow the hash character.**
- _MUST begin with # in first byte of record._
- _MUST contain all comment text within the line or be continued as a separate comment on additional lines._

**DataRec (required) = c1_val, c2, val, c3_val**
- _One or more records SHOULD be present but crap does happen sometimes._
- _MUST have a column value, even if null (,,), for all columns identified within the header rec._

I scanned the csvHelper code (quickly) and the documentation (somewhat) thoroughly and found the comment (#) character component, skip-empty lines and column count difference handling.  I need a csvHelper subclass or csvHelper itself to be able to handle the CSV format above and while I can hack a double-read operation to trim the version record and other @ records out, I am wondering if I can subclass some csvHelper class or set some properties to appropriately handle the above spec because I want/need to be able to use the mapping feature (reason I am actually here -- thx to tons of references from other developers/forums) in order to have column headers properly mapped to record class fields instead of just string array (indexing nightmare on revisions).  So, about 99% of the functionality I need is present in csvHelper.  Hoping that someone here has a suggestion that would allow me to consume this awesome piece of work (thx @JoshClose).

Just in case anyone asks "Why the @ records?":
The @ records are used to determine which revision of the CSV spec is to be used to parse the file since: columns could be added/deleted, revised as to how their value should be handled, and indicate whether columns of data are to be expected or not and how to handle accordingly during the consuming (my) application's logic operations (post-load/post-parse of this class library operations).

Thank you in advance.

**Comment by DarkonIV** [comment]
[csvfile a]
@ version 0.1.1
# system_id, owner, geohash
br549, bugs_bunny, 9yzy2f
omgi42, marvin_martian, 9yzy2f
# the following system is in for repair   <-- want to ignore processing (skip) this comment line
owdmdg, elmer_fudd, 9zp8kd

[csvfile b -- this file will only have these three records, ever, with CRLF at end of 3rd record]
@ version 0.2.5
# controller_id, controller_ver, controller_fw_ver, custom_fw_ver
47592837565, ABG 8M 6.24.17.161, 0.19.121.17, 0.0.11.1


[csvfile c]
@ version 1.2.1
# rec_id, event_id, data_id, lat, lon, date, time
1, 1, , 38.1764655, -90.4656644, 20180218, 131744
2, 1, 1, 38.1764655, -90.4656644, 20180218, 131744
3, 2, , 38.1764658, -90.4656641, 20180218, 131746
4, 2, 2, 38.1764658, -90.4656641, 20180218, 131746
1, 3, , 38.1764659, -90.4656641, 20180218, 131748
2, 3, 3, 38.1764659, -90.4656641, 20180218, 131748
3, 4, , 38.1764660, -90.4656641, 20180218, 131750
4, 4, 4, 38.1764660, -90.4656641, 20180218, 131750

IF we do not detect an @ version record, we default to mapping to a specific current-as-of-time-of-coding specification since some CSV files will not have been updated to this spec since that record would be the only change -- so those file formats, since haven't changed in many years and would break many systems, will not be updated at this time.  however, a new system i am working on uses CSV files everywhere (DB preload/init inject records, config recipes, operations data (like csvfile c above), and system identification files (like csvfile b above)).  There is speculation that other @ prefixed records could be augmentation indicators for HOW the values are to be interpreted and apply to all values in specific columns in order to NOT have to write the value repeatedly on each record.  A common example of this in geoprocessing is CRS (coordinate reference system; there are MANY; and each device can process internally on one CRS and output in another CRS where we only care about its output), where epsg 4326 would be on the @ augmentation record (maybe like @ geo:crs:4326 or maybe @ geo crs:4326) and we just take the : delimited data pieces and apply them accordingly during processing.  In some cases this value maybe be repeated and stored with each record which is needed or could be available as a global (maybe) of the entire data set/file and refer to internal processing paths that should be utilized and not an actual value of the record, so to speak.  But I am not overly concerned about any NEW, WHAT-IF scenarios of the @ records other than the @ version record for now.  But would like to have a pattern that could easily adopt new @ records as they may be devised.

in your final paragraph above, am i understanding that that GetRecords<T>() would read from the current record pointer location and not just re-slurp the entire file?  if so, then that would be a consideration albeit i have not yet understood how i would go about implementing this -- apologies for being such a dunce but my current projects schedules are likely stressing me to the point that i am not fully understanding the simplicity of what you may be suggesting.  Any pointers or example to kick-start my discovery would be sincerely appreciated.

thank you for replying and i look forward to investigating this seemingly very helpful and useful library.

**Comment by JoshClose** [comment]
Here is how you can read the example data you gave. Hopefully you can make changes to suit your needs.

I would suggest creating a class for each version that parser it. Figure out the version, then run it through that class. I'm just using methods in this example here.

```cs
void Main()
{
	var fileA = new StringBuilder();
	fileA.AppendLine("@ version 0.1.1");
	fileA.AppendLine("# system_id, owner, geohash");
	fileA.AppendLine("br549, bugs_bunny, 9yzy2f");
	fileA.AppendLine("omgi42, marvin_martian, 9yzy2f");
	fileA.AppendLine("# the following system is in for repair");
	fileA.AppendLine("owdmdg, elmer_fudd, 9zp8kd");

	var fileB = new StringBuilder();
	fileB.AppendLine("@ version 0.2.5");
	fileB.AppendLine("# controller_id, controller_ver, controller_fw_ver, custom_fw_ver");
	fileB.AppendLine("47592837565, ABG 8M 6.24.17.161, 0.19.121.17, 0.0.11.1");

	var fileC = new StringBuilder();
	fileC.AppendLine("@ version 1.2.1");
	fileC.AppendLine("# rec_id, event_id, data_id, lat, lon, date, time");
	fileC.AppendLine("1, 1, , 38.1764655, -90.4656644, 20180218, 131744");
	fileC.AppendLine("2, 1, 1, 38.1764655, -90.4656644, 20180218, 131744");
	fileC.AppendLine("3, 2, , 38.1764658, -90.4656641, 20180218, 131746");
	fileC.AppendLine("4, 2, 2, 38.1764658, -90.4656641, 20180218, 131746");
	fileC.AppendLine("1, 3, , 38.1764659, -90.4656641, 20180218, 131748");
	fileC.AppendLine("2, 3, 3, 38.1764659, -90.4656641, 20180218, 131748");
	fileC.AppendLine("3, 4, , 38.1764660, -90.4656641, 20180218, 131750");
	fileC.AppendLine("4, 4, 4, 38.1764660, -90.4656641, 20180218, 131750");

	Read(fileA.ToString());
	Read(fileB.ToString());
	Read(fileC.ToString());
}

public void Read(string file)
{
	using (var reader = new StringReader(file))
	using (var csv = new CsvReader(reader))
	{
		csv.Configuration.TrimOptions = TrimOptions.Trim;
		csv.Configuration.PrepareHeaderForMatch = header => header.Trim('#').Trim();
		csv.Configuration.AllowComments = false;
		csv.Configuration.RegisterClassMap<Version_0_1_1Map>();
		csv.Configuration.RegisterClassMap<Version_0_2_5Map>();
		csv.Configuration.RegisterClassMap<Version_1_2_1Map>();
		csv.Read();
		var version = Regex.Match(csv.GetField(0), @"@ version (?<Version>[0-9]\.[0-9]\.[0-9])").Groups["Version"]?.Value;
		csv.Read();
		csv.ReadHeader();
		csv.Configuration.AllowComments = true;
		switch (version)
		{
			case "0.1.1":
				csv.GetRecords<Version_0_1_1>().Dump();
				break;
			case "0.2.5":
				csv.GetRecords<Version_0_2_5>().Dump();
				break;
			case "1.2.1":
				csv.GetRecords<Version_1_2_1>().Dump();
				break;
		}
	}
}

public class Version_0_1_1
{
	public string SystemId { get; set; }
	public string Owner { get; set; }
	public string GeoHash { get; set; }
}

public class Version_0_1_1Map : ClassMap<Version_0_1_1>
{
	public Version_0_1_1Map()
	{
		Map(m => m.SystemId).Name("system_id");
		Map(m => m.Owner).Name("owner");
		Map(m => m.GeoHash).Name("geohash");
	}
}

public class Version_0_2_5
{
	public long ControllerId { get; set; }
	public string ControllerVer { get; set; }
	public string ControllerFwVer { get; set; }
	public string CustomFwVer { get; set; }
}

public class Version_0_2_5Map : ClassMap<Version_0_2_5>
{
	public Version_0_2_5Map()
	{
		Map(m => m.ControllerId).Name("controller_id");
		Map(m => m.ControllerVer).Name("controller_ver");
		Map(m => m.ControllerFwVer).Name("controller_fw_ver");
		Map(m => m.CustomFwVer).Name("custom_fw_ver");
	}
}

public class Version_1_2_1
{	
	public long RecId { get; set; }
	public long EventId { get; set; }
	public long? DataId { get; set; }
	public decimal Lat { get; set; }
	public decimal Lon { get; set; }
	public DateTime Date { get; set; }
	public string Time { get; set; }
}

public class Version_1_2_1Map : ClassMap<Version_1_2_1>
{
	public Version_1_2_1Map()
	{
		Map(m => m.RecId).Name("rec_id");
		Map(m => m.EventId).Name("event_id");
		Map(m => m .DataId).Name("data_id");
		Map(m => m.Lat).Name("lat");
		Map(m => m.Lon).Name("lon");
		Map(m => m.Date).Name("date").TypeConverterOption.Format("yyyyMMdd");
		Map(m => m.Time).Name("time");
	}
}
```

**Comment by DarkonIV** [comment]
Thank you @josh.  I will have time to look this over on Friday.  Thank you for the kicks starter code.

**Comment by DarkonIV** [comment]
Again, @JoshClose - thank you so much for the kick start code.  I have been building my file processor classes, csv file classes and testing - works like a charm with a few minor details that I was able to find a work around for and not dirty up the my architecture too much.  Yes, I realize that I could have just one processor to rule them all as well but since several of my files have the same version in them I opted to segregate the processing logic by specific class data realm.

_outline of what I have implemented without going into details (for you and anyone else):_

**_abstract CsvProcessorBase class -_** 
handles the common configuration, version extraction, header extraction, post header read re-config (optional), and common record parser (after mapped) which allows for one piece of code to track if record is ADD (no-tilde), DELETE (has-tilde), or is COMMENT (has # appearing as 1st char on DATA line within the CSV file.

**_CsvFormatAProcessor class_** : derived from CsvProcessorBase
handles all versions and special handling required for all VERSIONS of FileFormatA and includes the stream/csv readers enumerable function to be used within whatever class needs to read a FileFormatA of any version.
- in my case there will be one of these for each differently formatted CSV related to different TYPES of data records: (ex.) CsvStudentsProcessor, CsvJobsProcessor, CsvClassesProcessor, etc...
- However, there will only be this one class for handling reads for all VERSIONS of the CSV FILE TYPE so the Students processor class can handle "Students" files regardless if version 1.0 or 99999.9999.999.

**_abstract CsvVersionBase class -_**
handles and provides the common requirements of all Csv File Formats within this system:
- detection and proper field/property handling of DELETE tagged records;
- detection and proper field/property handling of COMMENT tagged records;
-- just to name a few common things all of these files have in common.

**_CsvFormatA_V_1_0_0 class_** - derived from CsvVersionBase
defines the properties/fields for Version 1.0.0 of File Format A (aka Students).
- Of course there will be a 1:1 **_Map_** class (CsvFormatA_V_1_0_0_Map) for each version of these classes which will be registered by the proper **_Processor_** class above.

**_CsvFormatA_V_1_5_0 class_** - TBD : derived from CsvVersionBase or CsvFormatA_V_1_0_0
- a revision of Format A class above regardless of how this class gets derived, it brings (for example) version 1.5 capability to the tooling while allowing BOTH/ALL versions of Format A to be read and handled.

If there is interest in this pattern, when I get some time in about a month, I could sterilize the pattern a bit and offer it up here or in snippet somewhere.

I am just so excited that this is falling into place so nicely -- thank you for providing such a wonderful piece of work.
